src/ApplicationInsights.ServiceFabric.Native/Net45/DependencyTrackingModule/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,324 @@
+        private readonly TimeSpan intervalBetweenFailedRetries = TimeSpan.FromSeconds(30);
+
+        private Uri endpointAddress;
+
+        private ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        private ConcurrentDictionary<string, int> fetchTasks = new ConcurrentDictionary<string, int>();
+
+        // Stores failed instrumentation keys along with the time we tried to retrieve them.
+        private ConcurrentDictionary<string, FailedResult> failingInstrumenationKeys = new ConcurrentDictionary<string, FailedResult>();
+
+        private Func<string, Task<string>> provideAppId;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationIdLookupHelper" /> class mostly to be used by the test classes to provide an override for fetching appId logic.
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId.</param>
+        public CorrelationIdLookupHelper(Func<string, Task<string>> appIdProviderMethod)
+        {
+            if (appIdProviderMethod == null)
+            {
+                throw new ArgumentNullException(nameof(appIdProviderMethod));
+            }
+
+            this.provideAppId = appIdProviderMethod;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationIdLookupHelper" /> class mostly to be used by the test classes to seed the instrumentation key -> app Id relationship.
+        /// </summary>
+        /// <param name="mapSeed">A dictionary that contains known instrumentation key - app id relationship.</param>
+        public CorrelationIdLookupHelper(Dictionary<string, string> mapSeed)
+        {
+            if (mapSeed == null)
+            {
+                throw new ArgumentNullException(nameof(mapSeed));
+            }
+
+            this.provideAppId = this.FetchAppIdFromService;
+
+            foreach (var entry in mapSeed)
+            {
+                this.knownCorrelationIds[entry.Key] = entry.Value;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationIdLookupHelper" /> class.
+        /// </summary>
+        /// <param name="endpointAddress">Endpoint that is to be used to fetch appId.</param>
+        public CorrelationIdLookupHelper(string endpointAddress)
+        {
+            if (string.IsNullOrEmpty(endpointAddress))
+            {
+                throw new ArgumentNullException(nameof(endpointAddress));
+            }
+
+            Uri endpointUri = new Uri(endpointAddress);
+
+            // Get the base URI, so that we can append the known relative segments to it.
+            this.endpointAddress = new Uri(endpointUri.AbsoluteUri.Substring(0, endpointUri.AbsoluteUri.Length - endpointUri.LocalPath.Length));
```

Comment:
Uri.GetLeftPath(UriPartial.Authority)?

Reply:
This is a copy paste of code from else where where it has been well tested. I'll keep it as it is.